### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by Nashtare**
# Description

Adds missing test scenarios to ensure edge cases are covered.

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

**Review by dvdplm — COMMENTED** [review]
test

File: src/parameters.rs [hunk 1]
```diff
@@ -62,4 +62,22 @@ mod test {
             assert_eq!(params, ThresholdParameters::from_bytes(&bytes).unwrap());
         }
     }
+
+    #[test]
+    #[should_panic]
+    fn test_no_participant() {
```
Comment (dvdplm) [inline src/parameters.rs:68]:
```suggestion
    fn requires_at_least_one_participant() {
```

File: src/parameters.rs [hunk 2]
```diff
@@ -62,4 +62,22 @@ mod test {
...
+        let _ = ThresholdParameters::<Secp256k1Sha256>::new(0, 0);
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_threshold_is_zero() {
```
Comment (dvdplm) [inline src/parameters.rs:74]:
```suggestion
    fn threshold_cannot_be_zero() {
```

File: src/keys.rs [hunk 3]
```diff
@@ -256,3 +256,63 @@ impl<C: CipherSuite> GroupVerifyingKey<C> {
...
+    use crate::utils::Scalar;
+    use crate::{FromBytes, ToBytes};
+    use ark_ff::UniformRand;
+    use rand::{rngs::OsRng, RngCore};
+
+    #[test]
+    fn test_serialization() {
+        let mut rng = OsRng;
+
+        for _ in 0..100 {
```
Comment (dvdplm) [inline src/keys.rs:273]: What is the purpose of doing this 100 times?

File: src/keys.rs [hunk 4]
```diff
@@ -256,3 +256,63 @@ impl<C: CipherSuite> GroupVerifyingKey<C> {
...
+    use super::*;
+    use crate::testing::Secp256k1Sha256;
+    use crate::utils::Scalar;
+    use crate::{FromBytes, ToBytes};
+    use ark_ff::UniformRand;
+    use rand::{rngs::OsRng, RngCore};
+
+    #[test]
+    fn test_serialization() {
+        let mut rng = OsRng;
```
Comment (dvdplm) [inline src/keys.rs:271]: DQ: is it true that users should **not** use `OsRng` in production, not unless they are positive that their OS uses a proper CSPRNG? (That may not be the case on an embedded device for instance)

File: src/keys.rs [hunk 5]
```diff
@@ -256,3 +256,63 @@ impl<C: CipherSuite> GroupVerifyingKey<C> {
...
+            );
+            let bytes = pkey.to_bytes().unwrap();
+            assert!(DiffieHellmanPublicKey::<Secp256k1Sha256>::from_bytes(&bytes).is_ok());
+            assert_eq!(
+                pkey,
+                DiffieHellmanPublicKey::<Secp256k1Sha256>::from_bytes(&bytes).unwrap()
+            );
+
+            let skey = IndividualSigningKey::<Secp256k1Sha256> {
+                index: rng.next_u32(),
```
Comment (dvdplm) [inline src/keys.rs:293]: A random comment, unrelated to this PR: is it ok to have `index` be a plain `u32`, as opposed to `std::num::Wrapping` or `std::num::Saturating`? Could we avoid allowing users to set the index? If we can't, can we make them use a setter you think? Not sure what the risk would be but it seems bad to allow unexpected wrap here.

File: src/keys.rs [hunk 6]
```diff
@@ -256,3 +256,63 @@ impl<C: CipherSuite> GroupVerifyingKey<C> {
...
+            let bytes = skey.to_bytes().unwrap();
+            assert!(IndividualSigningKey::<Secp256k1Sha256>::from_bytes(&bytes).is_ok());
+            assert_eq!(
+                skey,
+                IndividualSigningKey::<Secp256k1Sha256>::from_bytes(&bytes).unwrap()
+            );
+
+            let pkey = skey.to_public();
+            let bytes = pkey.to_bytes().unwrap();
+            assert!(IndividualVerifyingKey::<Secp256k1Sha256>::from_bytes(&bytes).is_ok());
```
Comment (dvdplm) [inline src/keys.rs:305]: Do we really need this assert, given what is coming after?

File: src/keys.rs [hunk 7]
```diff
@@ -256,3 +256,63 @@ impl<C: CipherSuite> GroupVerifyingKey<C> {
...
+            assert_eq!(
+                pkey,
+                IndividualVerifyingKey::<Secp256k1Sha256>::from_bytes(&bytes).unwrap()
+            );
+        }
+
+        let bytes = vec![255; 157];
```
Comment (dvdplm) [inline src/keys.rs:312]: Maybe it's obvious why this is an invalid key, but I don't get it. Either way: what do you think about moving this to its own test?

File: src/sign/signature.rs [hunk 8]
```diff
@@ -1371,6 +1371,79 @@ mod test {
...
+        let (p1_public_comshares, _) =
+            generate_commitment_share_lists::<Secp256k1Sha256>(&mut OsRng, &p1_sk, 1).unwrap();
+
+        let mut aggregator = SignatureAggregator::new(
+            params,
+            GroupVerifyingKey::new(Projective::zero()),
+            &message[..],
+        );
+
+        // The aggregator should panic as the participant index does not match the public key index.
```
Comment (dvdplm) [inline src/sign/signature.rs:1394]: A general comment on the library: is it really fair that a library crashes the whole process like this? I think that should be a decision for the app that uses the library. I guess there are catastrophic situations where the library *must* crash, but unless this is one of them, I think it would be better to return error.

File: src/sign/signature.rs [hunk 9]
```diff
@@ -1371,6 +1371,79 @@ mod test {
...
+            GroupVerifyingKey::new(Projective::zero()),
+            &message[..],
+        );
+
+        aggregator.include_signer(2, p2_public_comshares.commitments[0], &p2_sk.to_public());
+        aggregator.include_signer(1, p1_public_comshares.commitments[0], &p1_sk.to_public());
+        aggregator.include_signer(4, p4_public_comshares.commitments[0], &p4_sk.to_public());
+        aggregator.include_signer(3, p3_public_comshares.commitments[0], &p3_sk.to_public());
+        // Duplicates are fine, as we will remove them when gathering the signers list.
+        aggregator.include_signer(2, p2_public_comshares.commitments[0], &p2_sk.to_public());
```
Comment (dvdplm) [inline src/sign/signature.rs:1441]: This should be its own test, otherwise we would not notice if this behaviour wrt to duplicates ever changed. This test would still pass.

Using `#[should_panic]` is a bit dangerous like that: *any* panic will make the test pass...

File: src/dkg/key_generation.rs [hunk 10]
```diff
@@ -1292,6 +1292,311 @@ mod test {
...
+            p2_wrong_share,
+            p3_wrong_share,
+            p4_wrong_share,
+            p5_wrong_share,
+        ];
+
+        let result = p1_state.to_round_two(&p1_my_encrypted_secret_shares, rng);
+
+        assert!(result.is_err());
+        if let Error::Complaint(complaints) = result.unwrap_err() {
```
Comment (dvdplm) [inline src/dkg/key_generation.rs:1593]: Are the complaints ordered? Order of insertion or by index?

File: src/keys.rs (hunk 3 above)
Comment (Nashtare) [reply src/keys.rs:273]: I guess it's better to have it several times, and on random inputs, than a single time on fixed ones, no?

File: src/keys.rs (hunk 4 above)
Comment (Nashtare) [reply src/keys.rs:271]: Yeah for embedded devices you may have issues. I usually always use `OsRng` for testing as it's part of `rand_core`. Using `thread_rng` is good enough for seeding as well.

File: src/keys.rs (hunk 5 above)
Comment (Nashtare) [reply src/keys.rs:293]: The index of the signing key is equal to the index of the participant in the list.
That is, for a `t-out-of-n` scheme, the index will be `index <= n`.

I doubt that we need to have `n > 4B` and as such, we stuck with `u32` all around to save on some serialization size.

Also, I'm not sure why we'd get any saturating / wrapping operation here? We don't do any arithmetic operation on the index, it's just randomly generated here for the sake of testing serialization.

File: src/keys.rs (hunk 6 above)
Comment (Nashtare) [reply src/keys.rs:305]: fair, could be removed.

File: src/keys.rs (hunk 7 above)
Comment (Nashtare) [reply src/keys.rs:312]: I could add a comment on it, yes. The idea is that `[255; 157]` is just way too long (157 bytes, for a scalar field that is 32 bytes). (not sure where I picked this constant actually)

File: src/keys.rs (hunk 3 above)
Comment (dvdplm) [reply src/keys.rs:273]: Hmm, probably not! :)

Like: if there is an element of indeterminism and there could be a random key that causes the code to behave unexpectedly, then we should use a proper fuzzer and "do it properly". If there is not, well, then trying a 100 times changes nothing.

Why is 100 "enough" but 10 is not? Should it be 1000? Or 2^64 times?

File: src/keys.rs (hunk 5 above)
Comment (dvdplm) [reply src/keys.rs:293]: We are not doing any arithmetic here (or anywhere else afaik?) but it's a public value (ok, `pub(crate)`) so a clumsy dev *could* misuse it.

But yeah, now that I re-read this I don't think this concern is valid.

File: src/keys.rs (hunk 7 above)
Comment (dvdplm) [reply src/keys.rs:312]: aha, obviously! :)

Let's just rename the lvar from `bytes` to `too_long_key` or something?

Comment (dvdplm) [reply src/keys.rs:312]: I still think it should be its own test though.

File: src/keys.rs (hunk 3 above)
Comment (4l0n50) [reply src/keys.rs:273]: I agree it looks arbitrary the 100 times, but on the other hand it might not so simple to find the exact number. so why not 100 for now.

File: src/sign/signature.rs [hunk 11]
```diff
@@ -1371,6 +1371,79 @@ mod test {
...
+            index: 3,
+            key: Fr::rand(&mut OsRng),
+        };
+        let p4_sk = IndividualSigningKey {
+            index: 4,
+            key: Fr::rand(&mut OsRng),
+        };
+
+        let (p1_public_comshares, _) =
+            generate_commitment_share_lists::<Secp256k1Sha256>(&mut OsRng, &p1_sk, 1).unwrap();
```
Comment (4l0n50) [inline src/sign/signature.rs:1422]: I was confused because the name of this method doesn't mention the nonces. What about `generate_nonce_commitment_share_lists` (at least the paper says "nonce/commitment share'). Still, I don't see why the paper talks about "share" when you only have nonces (e_{ij}, d_{i,j}) and commitments to this nonces (D_{ij}, E_{ij})

Comment (4l0n50) [reply src/sign/signature.rs:1422]: And unrelated, why is E hidding? For preventing some attack?

**Review by 4l0n50 — APPROVED** [review]
LGTM
